moo.js

```diff
@@ -322,13 +322,25 @@
 
       // consume rest of buffer
       text = value = buffer.slice(index)
-      re.lastIndex = buffer.length
 
       // throw, if no rule with {error: true}
       if (!group) {
-        var index = text.indexOf('\n')
-        var firstLine = index === -1 ? text : text.slice(0, index)
-        throw new Error("Invalid syntax: line " + this.line + ": '" + firstLine + "'")
+        // seek to end
+        this.index = buffer.length
+
+        var start = Math.max(0, index - this.col + 1)
+        var eol = text.indexOf('\n')
+        if (eol === -1) eol = text.length
+        var line = buffer.slice(start, index + eol)
+        var message = ""
+        message += "invalid syntax at line " + this.line + " col " + this.col + ":\n\n"
+        message += "  " + line + "\n"
+        message += "  "
+        for (var i=this.col - 1; i--; ) {
+          message += " "
+        }
```

Comment:
I think this fails when `this.col > index + 1`, which could happen if you use `save()`. (Also, `Array(this.col).join(' ')` is a nice one-line way to do this if you like that sort of thing.)